### PR TITLE
define @timeout

### PR DESCRIPTION
**Pull request by felipenoris**
Quite useful macro.

~~Also, `doc/stdlib/linalg.rst` was altered by running `genstdlib.jl` script~~

### CODE REVIEW

File: base/event.jl [hunk 1]
```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
...
+
+"""
+    @set_timeout t ex [msg]
+
+Set a timeout of `t` seconds on the execution time of expression `ex`.
+Throws `TimeoutException` with optional error message `msg`
+if the timeout is triggered.
+"""
+macro set_timeout(timeout_s, ex, msgs...)
+    t_ = eval(timeout_s)
```
Comment (TotalVerb) [inline base/event.jl]: why evaluate this at expansion time instead of at runtime?

File: base/event.jl [hunk 2]
```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
...
+
+    msg = isempty(msgs) ? ex : msgs[1]
```
Comment (TotalVerb) [inline base/event.jl]: Don't accept multiple `msgs` if only the first will be used. Macros are generic now, so default argument/creating another method should work.

File: base/event.jl [hunk 3]
```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
...
+    if !isempty(msgs) && (isa(msg, Expr) || isa(msg, Symbol))
+        # message is an expression needing evaluating
+        msg = :(Main.Base.string($(esc(msg))))
+    elseif isdefined(Main, :Base) && isdefined(Main.Base, :string)
```
Comment (TotalVerb) [inline base/event.jl]: Is this special case needed? Overhead of `string("string literal")` is extremely small.

File: base/event.jl [hunk 4]
```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
...
+        msg = Main.Base.string(msg)
+    else
+        # string() might not be defined during bootstrap
+        msg = :(Main.Base.string($(Expr(:quote,msg))))
```
Comment (TotalVerb) [inline base/event.jl]: I don't think `:quote` is needed here. But see above.

File: base/event.jl [hunk 5]
```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
...
+    end
+
+    quote
+        let
+            wt = Condition()
+            timeout_occurred = true
+
+            @schedule ( try r = $(esc(ex)); timeout_occurred = false; notify(wt, r); catch e; notify_error(wt, e); end )
```
Comment (TotalVerb) [inline base/event.jl]: I think this would look better as

```
@schedule try
    ...
end
```

File: base/event.jl [hunk 6]
```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
...
+            @schedule ( sleep($t_); notify(wt) )
+
+            result = wait(wt)
+            timeout_occurred && throw(TimeoutException($msg))
+
+            result
+            end
```
Comment (TotalVerb) [inline base/event.jl]: whitespace

File: doc/stdlib/linalg.rst [hunk 7]
```diff
@@ -322,7 +322,7 @@ Linear algebra functions in Julia are largely implemented by calling functions f
...
+   Compute the Cholesky factorization of a sparse positive definite matrix ``A``\ . ``A`` must be a ``SparseMatrixCSC``\ , ``Symmetric{SparseMatrixCSC}``\ , or ``Hermitian{SparseMatrixCSC}``\ . Note that even if ``A`` doesn't have the type tag, its structure and values must still be symmetric/Hermitian. A fill-reducing permutation is used. ``F = cholfact(A)`` is most frequently used to solve systems of equations with ``F\b``\ , but also the methods ``diag``\ , ``det``\ , ``logdet`` are defined for ``F``\ . You can also extract individual factors from ``F``\ , using ``F[:L]``\ . However, since pivoting is on by default, the factorization is internally represented as ``A == P'*L*L'*P`` with a permutation matrix ``P``\ ; using just ``L`` without accounting for ``P`` will give incorrect answers. To include the effects of permutation, it's typically preferable to extact "combined" factors like ``PtL = F[:PtL]`` (the equivalent of ``P'*L``\ ) and ``LtP = F[:UP]`` (the equivalent of ``L'*P``\ ).
```
Comment (TotalVerb) [inline doc/stdlib/linalg.rst]: Are these changes related?

Comment (felipenoris) [inline doc/stdlib/linalg.rst]: As I mentioned, this is generated automatically. In order to update the docs I must run `genstdlib.jl`, which updates all the docs from docstrings.

File: base/event.jl (hunk 5 above)
Comment (felipenoris) [inline base/event.jl]: I agree.

File: base/event.jl (hunk 3 above)
Comment (felipenoris) [inline base/event.jl]: This part of the code was based on `macro assert(ex, msgs...)` defined in `error.jl`.

File: base/event.jl (hunk 2 above)
Comment (felipenoris) [inline base/event.jl]: Cool. I'll take a look. This was based on `macro assert(ex, msgs...)` defined in `error.jl`.

File: doc/stdlib/linalg.rst (hunk 7 above)
Comment (TotalVerb) [inline doc/stdlib/linalg.rst]: Yes, sorry, I had missed that.

File: test/event.jl [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+# Tests for @set_timeout
+event_const_a = 10
+event_fun_x(a) = ( sleep(0.1); 100a )
+event_fun_xa(a) = 100a
+event_fun_y() = ( sleep(0.1); 200 )
+event_fun_w(a, b_) = a + b_
+event_fun_w2(a, b_) = (sleep(0.1); a + b_)
+
+function event_fun_z()
+   	sleep(0.1)
```
Comment (tkelman) [inline test/event.jl]: 4-space indent

File: base/exports.jl [hunk 9]
```diff
@@ -168,6 +168,7 @@ export
     ProcessExitedException,
     RemoteException,
     SystemError,
+    TimeoutException,
```
Comment (tkelman) [inline base/exports.jl]: there's a table of exception types in the docs that this should be added to

File: base/event.jl (hunk 3 above)
Comment (tkelman) [inline base/event.jl]: that's most likely earlier in bootstrap and not needed here

**Comment by tkelman** [comment]
I wouldn't have the `set_` part in this name. Does this need to be a macro, or could it be simple enough to just use a do-block?

**Comment by yuyichao** [comment]
Also worth mentioning that this does not guarantee that the operation will be aborted when the time limit is reached and that raising an error doesn't mean the operation is aborted.

**Comment by felipenoris** [comment]
So, thanks for all your kind comments. I think I've gone thru all of them.
But still I'm having troubles with the two last test cases, which are failing:
```julia
@test_throws TimeoutException @timeout(0.01, event_fun_w2(1, event_const_a), "error " * "message")
@test_throws TimeoutException @timeout(0.01, event_fun_w2(1, event_const_a), event_msg)
```

**Comment by felipenoris** [comment]
Thanks for all your kind comments. I think I've gone thru all of them.

**Comment by felipenoris** [comment]
On @tkelman's idea, I'm open for your suggestions.

If I got it right, this does the thing:
```julia
type TimeoutException <: Exception
    msg::AbstractString

    TimeoutException() = new("")
    TimeoutException(msg) = new(msg)
end

function timeout(f, timeout_s, msg="", args...)
    wt = Condition()
    timeout_ocurred = true

    @schedule try
        r = f(args...)
        timeout_ocurred = false
        notify(wt, r)
    catch e
        notify(wt, e)
    end

    @schedule ( sleep(timeout_s); notify(wt) )

    result = wait(wt)
    timeout_ocurred && throw(TimeoutException(msg))
    result
end
```

So:

```julia
julia> timeout(1) do
           sleep(0.5)
           println("hello")
       end
hello

julia> timeout(1) do
           sleep(2)
           println("hello")
       end
ERROR: TimeoutException("")
 in timeout(::Function, ::Int64, ::String) at ./REPL[3]:16
 in timeout(::Function, ::Int64) at ./REPL[3]:2
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```

Is that your suggestion? I'm not sure if one is better the the other (do-block vs macro). I just liked the macro syntax better. Maybe you have a clearer view.

**Comment by kshyatt** [comment]
Needs a rebase.

File: base/event.jl [hunk 10]
```diff
@@ -343,3 +343,50 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
...
+
+"""
+    @timeout(t, ex, [msg])
+
+Set a timeout of `t` seconds on the execution time of expression `ex`.
+Throws `TimeoutException` with optional error message `msg`
+if the timeout is triggered.
+
+Be aware that the exception raised by the timeout will not abort
+the evaluation of `ex`.
```
Comment (yuyichao) [inline base/event.jl]: This still doesn't mention that the exception can be raised arbitrarily long after the specified timeout.

Comment (felipenoris) [inline base/event.jl]: @yuyichao, I don't know if I understand that correctly. Is that due to the "single-threaded" scheduler that might be doing other stuff before starting the `sleep(timeout)` task?

Comment (yuyichao) [inline base/event.jl]: No, It can simply be caused by the user function not hitting a `yield` at all.
This is also why I'm wondering how useful this function is.
I believe a timeout is usually used to make blocking code relatively non-blocking and to abort operation that is taking too long. This macro doesn't seems to provide any of these.

1. It can only provide feedback (of whether there's a timeout) for operations that are already async (i.e. it has to yield to other tasks)
2. It does not abort any operations.

(1) basically can't be solved with tasks.
For (2), it would be more useful if `sleep(timeout)` task runs a user defined callback instead, which at least has some chance to be able to abort the operation. However, this would remove the majority of the code in this PR and I feel like a `@schedule` with an optional delay might be more useful....

Comment (felipenoris) [inline base/event.jl]: Ok, looks like this would be done in an appropriate way in a future development.
